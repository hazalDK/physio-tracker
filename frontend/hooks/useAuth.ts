import * as SecureStore from "expo-secure-store";
import axios from "axios";
import { useNavigation } from "@react-navigation/native";
import { StackNavigationProp } from "@react-navigation/stack";
import { RootStackParamsList } from "@/types/navigation";
import { Alert } from "react-native";

export function useAuth() {
  const navigation =
    useNavigation<StackNavigationProp<RootStackParamsList, "login">>();

  const getToken = async () => {
    return await SecureStore.getItemAsync("access_token");
  };

  const refreshToken = async () => {
    try {
      const refreshToken = await SecureStore.getItemAsync("refresh_token");
      if (!refreshToken) {
        throw new Error("No refresh token available");
      }

      const refreshUrl = process.env.API_URL || "http://192.168.68.111:8000";
      const refreshResponse = await axios.post(
        `${refreshUrl}/api/token/refresh/`,
        { refresh: refreshToken }
      );

      const newToken = refreshResponse.data.access;
      const newRefreshToken = refreshResponse.data.refresh;

      await Promise.all([
        SecureStore.setItemAsync("access_token", newToken),
        SecureStore.setItemAsync("refresh_token", newRefreshToken),
      ]);

      return newToken;
    } catch (error) {
      await Promise.all([
        SecureStore.deleteItemAsync("access_token"),
        SecureStore.deleteItemAsync("refresh_token"),
      ]);

      Alert.alert("Session Expired", "Please login again", [
        { text: "OK", onPress: () => navigation.navigate("login") },
      ]);

      return null;
    }
  };

  const createApiInstance = async () => {
    let token = await getToken();

    if (!token) {
      Alert.alert("Login Required", "Please sign in to continue", [
        { text: "OK", onPress: () => navigation.navigate("login") },
      ]);
      return null;
    }

    return axios.create({
      baseURL: process.env.API_URL || "http://192.168.68.111:8000",
      timeout: 10000,
      headers: {
        Authorization: `Bearer ${token}`,
        "Content-Type": "application/json",
      },
    });
  };

  return { getToken, refreshToken, createApiInstance };
}
