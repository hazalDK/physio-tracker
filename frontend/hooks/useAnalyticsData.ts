import { useState, useEffect } from "react";
import axios from "axios";
import { useAuth } from "./useAuth";

// Helper function to format date
const formatDate = (date: Date) => {
  return date.toISOString().split("T")[0];
};

// Type for analytics data (can be either adherence or pain data)
interface ChartData {
  labels: string[];
  datasets: {
    data: number[];
  }[];
}

/**
 * Custom hook to fetch and manage analytics data (adherence or pain levels).
 * @param {string} endpoint - The API endpoint to fetch data from.
 * @param {"adherence" | "pain"} dataType - The type of data to fetch (either adherence or pain).
 * @returns {Object} - An object containing loading state, chart data, average, history, error, and functions to navigate weeks.
 * @throws {Error} - Throws an error if the API call fails.
 * */
export function useAnalyticsData<T>(
  endpoint: string,
  dataType: "adherence" | "pain"
): object {
  const [loading, setLoading] = useState(true);
  const [chartData, setChartData] = useState<ChartData>({
    labels: ["Mon", "Tue", "Wed", "Thu", "Fri", "Sat", "Sun"],
    datasets: [{ data: [0, 0, 0, 0, 0, 0, 0] }],
  });
  const [average, setAverage] = useState(0);
  const [history, setHistory] = useState<T[]>([]);
  const [error, setError] = useState<string | null>(null);

  // Add week navigation state
  const [currentEndDate, setCurrentEndDate] = useState(new Date());
  const [weekTitle, setWeekTitle] = useState("Current Week");
  const { createApiInstance, refreshToken } = useAuth();

  // Function to go to previous week
  const goToPreviousWeek = () => {
    const newEndDate = new Date(currentEndDate);
    newEndDate.setDate(newEndDate.getDate() - 7);
    setCurrentEndDate(newEndDate);
    fetchData(newEndDate);
    updateWeekTitle(newEndDate);
  };

  // Function to go to next week (but not beyond current date)
  const goToNextWeek = () => {
    const today = new Date();
    const newEndDate = new Date(currentEndDate);
    newEndDate.setDate(newEndDate.getDate() + 7);

    // Don't go beyond current date
    if (newEndDate > today) {
      newEndDate.setTime(today.getTime());
    }

    setCurrentEndDate(newEndDate);
    fetchData(newEndDate);
    updateWeekTitle(newEndDate);
  };

  // Update week title based on date range
  const updateWeekTitle = (endDate: Date) => {
    const startDate = new Date(endDate);
    startDate.setDate(startDate.getDate() - 6);

    const today = new Date();
    const isCurrentWeek =
      today.getDate() === endDate.getDate() &&
      today.getMonth() === endDate.getMonth() &&
      today.getFullYear() === endDate.getFullYear();

    if (isCurrentWeek) {
      setWeekTitle("Current Week");
    } else {
      const startMonth = startDate.toLocaleString("default", {
        month: "short",
      });
      const endMonth = endDate.toLocaleString("default", { month: "short" });

      if (startMonth === endMonth) {
        setWeekTitle(
          `${startMonth} ${startDate.getDate()} - ${endDate.getDate()}`
        );
      } else {
        setWeekTitle(
          `${startMonth} ${startDate.getDate()} - ${endMonth} ${endDate.getDate()}`
        );
      }
    }
  };

  // Function to fetch data for a specific end date
  const fetchData = async (endDate: Date) => {
    setLoading(true);
    try {
      const api = await createApiInstance();
      if (!api) return;

      try {
        // Fetch stats from the API with date parameter
        const response = await api.get(
          `${endpoint}?end_date=${formatDate(endDate)}`
        );
        if (response.data) {
          setChartData(response.data.chart_data);

          if (dataType === "adherence") {
            setAverage(response.data.average_adherence);
          } else {
            setAverage(response.data.average_pain);
          }

          setHistory(response.data.history);
        }
      } catch (error) {
        // Handle 401 error by refreshing the token and retrying the request
        if (axios.isAxiosError(error) && error.response?.status === 401) {
          // Attempt token refresh
          const newToken = await refreshToken();
          if (!newToken) return;

          const api = await createApiInstance();
          if (!api) return;

          // Retry with new token
          const response = await api.get(
            `${endpoint}?end_date=${formatDate(endDate)}`
          );

          if (response.data) {
            setChartData(response.data.chart_data);

            if (dataType === "adherence") {
              setAverage(response.data.average_adherence);
            } else {
              setAverage(response.data.average_pain);
            }

            setHistory(response.data.history);
          }
        } else {
          // Handle other errors
          console.error("API request failed:", error);
          setError(
            `Failed to load your ${
              dataType === "adherence" ? "exercises" : "pain data"
            }. Please check your connection and try again.`
          );
        }
      }
    } catch (error) {
      console.error(
        `Failed to load your ${
          dataType === "adherence" ? "exercises" : "pain data"
        }.`
      );
    } finally {
      setLoading(false);
    }
  };

  useEffect(() => {
    // Initial data fetch for current week
    fetchData(currentEndDate);
    updateWeekTitle(currentEndDate);
  }, []);

  return {
    loading,
    chartData,
    average,
    history,
    error,
    weekTitle,
    goToPreviousWeek,
    goToNextWeek,
  };
}
