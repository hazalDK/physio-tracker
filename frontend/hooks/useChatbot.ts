import { useState, useEffect, useCallback } from "react";
import { Alert } from "react-native";
import Toast from "react-native-root-toast";
import { messageType } from "@/types/chatbot";
import { createChatApi, apiRequestWithRefresh } from "../api";
import { useAuth } from "./useAuth";

export default function useChatbot() {
  const auth = useAuth();
  const [messages, setMessages] = useState<messageType[]>([]);
  const [userExercises, setUserExercises] = useState([]);
  const [recentPain, setRecentPain] = useState(0);
  const [isLoading, setIsLoading] = useState(false);
  const [isInitialLoading, setIsInitialLoading] = useState(true);
  const [api, setApi] = useState<any>(null);

  // Initialize API and chat
  useEffect(() => {
    const initializeApi = async () => {
      const apiInstance = await auth.createApiInstance();
      setApi(apiInstance);
    };

    initializeApi();

    // Set initial messages
    setMessages([
      {
        id: "disclaimer",
        text: "DISCLAIMER",
        sender: "disclaimer",
      },
      {
        id: "welcome",
        text: "Welcome to your physiotherapy assistant! How can I help you today?",
        sender: "bot",
      },
    ]);
  }, []);

  // Fetch user data when API is available
  useEffect(() => {
    if (api) {
      fetchUserData();
    }
  }, [api]);

  const fetchUserData = async () => {
    if (!api) return;

    setIsInitialLoading(true);
    const chatApi = createChatApi(api);

    try {
      // Use the wrapper function for API calls with refresh
      const fetchData = async () => {
        const [reportsData, exercisesData] = await Promise.all([
          chatApi.fetchReports(),
          chatApi.fetchUserExercises(),
        ]);

        return { reportsData, exercisesData };
      };

      const { reportsData, exercisesData } = await apiRequestWithRefresh(
        fetchData,
        auth
      );

      setUserExercises(exercisesData);

      if (reportsData.length > 0) {
        setRecentPain(reportsData[0].pain_level);
      }
    } catch (error) {
      console.error("Failed to fetch user data:", error);

      Alert.alert(
        "Error",
        "Failed to load your exercises. Please check your connection and try again."
      );
    } finally {
      setIsInitialLoading(false);
    }
  };

  const sendMessage = async (inputText: string) => {
    if (!inputText.trim() || !api) return;

    // Add user message to chat
    const userMessageId = Date.now().toString();
    setMessages((prev) => [
      ...prev,
      { id: userMessageId, text: inputText, sender: "user" },
    ]);

    setIsLoading(true);
    const chatApi = createChatApi(api);

    try {
      // Create context object with user's exercise data
      const exerciseContext = {
        exercises: userExercises,
        recentPain: recentPain,
      };

      // Use the wrapper function for API calls with refresh
      const sendMessageCall = async () => {
        return await chatApi.sendChatMessage(inputText, exerciseContext);
      };

      // Send message to API with refresh handling
      const response = await apiRequestWithRefresh(sendMessageCall, auth);

      // Add bot response to chat
      setMessages((prev) => [
        ...prev,
        {
          id: Date.now().toString(),
          text: response.message,
          sender: "bot",
        },
      ]);
    } catch (error) {
      console.error("Error sending message:", error);

      setMessages((prev) => [
        ...prev,
        {
          id: Date.now().toString(),
          text: "Sorry, I encountered an error. Please try again.",
          sender: "bot",
        },
      ]);
    } finally {
      setIsLoading(false);
    }
  };

  const resetChat = async () => {
    if (!api) return;

    const chatApi = createChatApi(api);

    try {
      // Use the wrapper function for API calls with refresh
      const resetCall = async () => {
        await chatApi.resetChat();
      };

      await apiRequestWithRefresh(resetCall, auth);

      // Reset messages
      setMessages([
        {
          id: "disclaimer",
          text: "DISCLAIMER",
          sender: "disclaimer",
        },
        {
          id: "welcome",
          text: "Welcome to your physiotherapy assistant! How can I help you today?",
          sender: "bot",
        },
      ]);

      // Refresh user data
      fetchUserData();

      Toast.show("Chat reset!", {
        duration: Toast.durations.SHORT,
        position: Toast.positions.BOTTOM,
        backgroundColor: "#14b8a6",
        textColor: "white",
        shadow: true,
      });
    } catch (error) {
      console.error("Failed to reset chat:", error);
      Alert.alert("Error", "Could not reset chat.");
    }
  };

  return {
    messages,
    isLoading,
    isInitialLoading,
    sendMessage,
    resetChat,
  };
}
