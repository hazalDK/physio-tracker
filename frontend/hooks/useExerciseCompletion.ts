import { useState } from "react";
import { Alert } from "react-native";
import { useNavigation } from "@react-navigation/native";
import { useAuth } from "./useAuth";

// Custom hook to manage user profile data and related operations.
export function useExerciseCompletion(userExerciseId: number | null) {
  const [showCompletionForm, setShowCompletionForm] = useState(false);
  const [reps, setReps] = useState(0);
  const [sets, setSets] = useState(0);
  const [painLevel, setPainLevel] = useState(0);
  const navigation = useNavigation();
  const { createApiInstance } = useAuth();

  // Function to handle saving exercise completion data
  // This function is called when the user submits the completion form.
  const handleSave = async () => {
    // Validate input fields
    if (!reps || !sets || !painLevel) {
      Alert.alert("Error", "Please fill all required fields");
      return;
    }

    // Api instance creation
    const api = await createApiInstance();
    if (!api) return;

    try {
      // Update user exercise data
      await api.put(`/user-exercises/${userExerciseId}/`, {
        reps: reps,
        sets: sets,
        pain_level: painLevel,
        completed: true,
      });

      // Closes the modal
      setShowCompletionForm(false);

      // Handle pain level logic
      // If pain level is low (1-3), check if they can increase difficulty
      if (painLevel <= 3) {
        try {
          const canIncreaseResponse = await api.get(
            `/user-exercises/${userExerciseId}/can_increase/`
          );

          if (canIncreaseResponse.data.can_increase) {
            Alert.alert(
              "Exercise Progress",
              "Your pain level is low and you've had consistent low pain for 3 days. Would you like to increase the difficulty for next time?",
              [
                {
                  text: "No",
                  onPress: () => navigation.goBack(),
                  style: "cancel",
                },
                {
                  text: "Yes",
                  onPress: async () => {
                    try {
                      await api.post(
                        `/user-exercises/${userExerciseId}/confirm_increase/`,
                        { confirm: "yes" }
                      );
                      Alert.alert(
                        "Success",
                        "Exercise difficulty increased for next time!"
                      );
                      navigation.goBack();
                    } catch (error) {
                      console.error("Error increasing difficulty:", error);
                      Alert.alert(
                        "Error",
                        "Failed to increase difficulty. Please try again."
                      );
                      navigation.goBack();
                    }
                  },
                },
              ]
            );
          } else {
            Alert.alert(
              "Exercise Progress",
              "Your pain level is low. Keep it up for 3 days to unlock higher difficulty!",
              [{ text: "OK", onPress: () => navigation.goBack() }]
            );
          }
        } catch (error) {
          // Handle error when checking increase eligibility
          console.error("Error checking increase eligibility:", error);
          Alert.alert("Success", "Exercise completed successfully", [
            { text: "OK", onPress: () => navigation.goBack() },
          ]);
        }
      } else if (painLevel > 3) {
        // If pain level is high, ask if they want to decrease difficulty
        Alert.alert(
          "Exercise Difficulty",
          "Your pain level is high. Would you like to decrease the difficulty for next time?",
          [
            {
              text: "No",
              onPress: () => navigation.goBack(),
              style: "cancel",
            },
            {
              text: "Yes",
              onPress: async () => {
                try {
                  await api.post(
                    `/user-exercises/${userExerciseId}/confirm_decrease/`,
                    { confirm: "yes" }
                  );
                  Alert.alert(
                    "Success",
                    "Exercise difficulty decreased for next time!"
                  );
                  navigation.goBack();
                } catch (error) {
                  console.error("Error decreasing difficulty:", error);
                  Alert.alert(
                    "Error",
                    "Failed to decrease difficulty. Please try again."
                  );
                  navigation.goBack();
                }
              },
            },
          ]
        );
      } else {
        Alert.alert("Success", "Exercise completed successfully", [
          { text: "OK", onPress: () => navigation.goBack() },
        ]);
      }
    } catch (error) {
      // Handle error when saving exercise completion
      console.error("Error saving exercise:", error);
      Alert.alert(
        "Error",
        "Failed to save exercise completion. Please try again."
      );
    }
  };

  // Generate dropdown data
  const getDropdownData = (maxValue: number) => {
    return Array.from({ length: maxValue }, (_, i) => ({
      label: `${i + 1}`,
      value: i + 1,
    }));
  };

  // Generate pain level dropdown data
  const getPainLevelData = () => {
    return Array.from({ length: 11 }, (_, i) => ({
      label: `${i}`,
      value: i,
    }));
  };

  return {
    showCompletionForm,
    setShowCompletionForm,
    reps,
    setReps,
    sets,
    setSets,
    painLevel,
    setPainLevel,
    handleSave,
    getDropdownData,
    getPainLevelData,
  };
}
