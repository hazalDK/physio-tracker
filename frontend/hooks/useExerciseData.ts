import { useState, useEffect } from "react";
import { Alert } from "react-native";
import axios from "axios";
import { ExerciseItem, UserExerciseItem } from "@/types/exercise";
import { useAuth } from "./useAuth";

export function useExerciseData(
  exerciseId: number,
  userExerciseId: number | null
) {
  const [loading, setLoading] = useState(true);
  const [exercise, setExercise] = useState<ExerciseItem>();
  const [userExercise, setUserExercise] = useState<UserExerciseItem>();
  const { createApiInstance, refreshToken } = useAuth();

  useEffect(() => {
    const fetchData = async () => {
      setLoading(true);
      try {
        const api = await createApiInstance();
        if (!api) return;

        try {
          const [userExercisesResponse, exercisesResponse] = await Promise.all([
            api.get(`/user-exercises/${userExerciseId}/`),
            api.get(`/exercises/${exerciseId}/`),
          ]);

          setUserExercise(
            !Array.isArray(userExercisesResponse.data)
              ? userExercisesResponse.data
              : ({} as UserExerciseItem)
          );

          setExercise(
            !Array.isArray(exercisesResponse.data)
              ? exercisesResponse.data
              : ({} as ExerciseItem)
          );
        } catch (error) {
          if (axios.isAxiosError(error) && error.response?.status === 401) {
            const newToken = await refreshToken();
            if (!newToken) return;

            const api = await createApiInstance();
            if (!api) return;

            const [retryUserExercises, retryExercises] = await Promise.all([
              api.get(`/user-exercises/${userExerciseId}/`),
              api.get(`/exercises/${exerciseId}/`),
            ]);

            setUserExercise(
              !Array.isArray(retryUserExercises.data)
                ? retryUserExercises.data
                : ({} as UserExerciseItem)
            );

            setExercise(
              !Array.isArray(retryExercises.data)
                ? retryExercises.data
                : ({} as ExerciseItem)
            );
          } else {
            console.error("API request failed:", error);
            Alert.alert(
              "Error",
              "Failed to load your exercises. Please check your connection and try again."
            );
            setExercise({} as ExerciseItem);
            setUserExercise({} as UserExerciseItem);
          }
        }
      } catch (error) {
        console.error("Failed to load exercise data:", error);
      } finally {
        setLoading(false);
      }
    };

    fetchData();
  }, [exerciseId, userExerciseId]);

  return { loading, exercise, userExercise };
}
