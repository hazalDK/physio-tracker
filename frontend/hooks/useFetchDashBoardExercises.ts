import { useState, useCallback } from "react";
import { Alert } from "react-native";
import axios from "axios";
import { ExerciseItem, UserExerciseItem } from "@/types/exercise";
import { useAuth } from "./useAuth";

/**
 * Custom hook to fetch and manage active exercises and user exercises data.
 */
export function useFetchDashboardExercises(): {
  loading: boolean;
  userExercises: UserExerciseItem[];
  exercises: ExerciseItem[];
  inActiveExercises: ExerciseItem[];
  refreshKey: number;
  fetchData: () => Promise<void>;
  refreshData: () => void;
} {
  const [loading, setLoading] = useState(true);
  const [userExercises, setUserExercises] = useState<UserExerciseItem[]>([]);
  const [exercises, setExercises] = useState<ExerciseItem[]>([]);
  const [inActiveExercises, setInActiveExercises] = useState<ExerciseItem[]>(
    []
  );
  const [refreshKey, setRefreshKey] = useState(0);
  const { createApiInstance, refreshToken } = useAuth();

  // Function to fetch active exercises and user exercises data and handle token refresh if needed
  const fetchData = useCallback(async () => {
    setLoading(true);

    try {
      // Create API instance
      const api = await createApiInstance();
      if (!api) return;

      try {
        // Fetch user exercises and active exercises data
        const [
          userExercisesResponse,
          exercisesResponse,
          inactiveExercisesResponse,
        ] = await Promise.all([
          api.get("/user-exercises/"),
          api.get("/users/active_exercises/"),
          api.get("/users/inactive_exercises/"),
        ]);

        // Set user exercises and active exercises data
        setUserExercises(
          Array.isArray(userExercisesResponse.data)
            ? userExercisesResponse.data
            : []
        );

        setExercises(
          Array.isArray(exercisesResponse.data) ? exercisesResponse.data : []
        );

        setInActiveExercises(
          Array.isArray(inactiveExercisesResponse.data)
            ? inactiveExercisesResponse.data
            : []
        );

        setInActiveExercises(
          Array.isArray(inactiveExercisesResponse.data)
            ? inactiveExercisesResponse.data
            : []
        );
      } catch (error) {
        // Handle 401 error by refreshing the token and retrying the request
        if (axios.isAxiosError(error) && error.response?.status === 401) {
          const newToken = await refreshToken();
          if (!newToken) return;

          // Create a new API instance with the refreshed token
          const api = await createApiInstance();
          if (!api) return;

          // Retry fetching user exercises and active exercises data
          const [retryUserExercises, retryExercises, retryInactiveExercises] =
            await Promise.all([
              api.get("/user-exercises/"),
              api.get("/users/active_exercises/"),
              api.get("/users/inactive_exercises/"),
            ]);

          // Set user exercises and active exercises data
          setUserExercises(
            Array.isArray(retryUserExercises.data)
              ? retryUserExercises.data
              : []
          );

          setExercises(
            Array.isArray(retryExercises.data) ? retryExercises.data : []
          );

          setInActiveExercises(
            Array.isArray(retryInactiveExercises.data)
              ? retryInactiveExercises.data
              : []
          );
        } else {
          // Handle other errors
          // Log the error and show an alert to the user
          console.error("API request failed:", error);
          Alert.alert(
            "Error",
            "Failed to load your exercises. Please check your connection and try again."
          );
          setExercises([]);
          setUserExercises([]);
        }
      }
    } catch (error) {
      console.error("Failed to load exercises:", error);
    } finally {
      setLoading(false);
    }
  }, []);

  // Function to refresh data by calling fetchData and updating the refresh key
  const refreshData = useCallback(() => {
    fetchData();
    setRefreshKey((prev) => prev + 1);
  }, [fetchData]);

  return {
    loading,
    userExercises,
    exercises,
    inActiveExercises,
    refreshKey,
    fetchData,
    refreshData,
  };
}
