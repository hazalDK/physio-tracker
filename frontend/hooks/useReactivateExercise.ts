import { useState } from "react";
import { useAuth } from "./useAuth";
import { Alert } from "react-native";

export function useReactivateExercise() {
  const [loading, setLoading] = useState(false);
  const { createApiInstance, refreshToken } = useAuth();

  // Function to reactivate an exercise
  const reactivateExercise = async (exerciseId: number) => {
    setLoading(true);
    try {
      // Create API instance
      const api = await createApiInstance();
      if (!api) {
        setLoading(false);
        return;
      }

      try {
        // Reactivate the exercise
        const response = await api.put(
          `/user-exercises/${exerciseId}/reactivate_exercise/`,
          { exercise_id: exerciseId }
        );

        if (response.status === 200 || response.status === 201) {
          Alert.alert("Success", "Exercise added successfully");
        } else {
          Alert.alert("Error", "Failed to add the exercise");
        }
      } catch (error: any) {
        // Handle 401 error by refreshing the token and retrying the request
        if (error.response?.status === 401) {
          // Attempt token refresh
          const newToken = await refreshToken();

          if (newToken) {
            // Create a new API instance with the refreshed token
            const api = await createApiInstance();
            if (!api) {
              setLoading(false);
              return;
            }

            try {
              const response = await api.put(
                `/user-exercises/${exerciseId}/reactivate_exercise/`,
                { exercise_id: exerciseId }
              );

              if (response.status === 200 || response.status === 201) {
                Alert.alert("Success", "Exercise reactivated successfully");
              } else {
                Alert.alert("Error", "Failed to reactivate the exercise");
              }
            } catch (secondError) {
              handleApiError(secondError);
            }
          } else {
            Alert.alert(
              "Error",
              "Failed to reactivate the exercise. Please check your connection and try again."
            );
          }
        } else {
          handleApiError(error);
        }
      }
    } catch (error: any) {
      console.error("API request failed:", error);
      Alert.alert(
        "Error",
        "Failed to reactivate the exercise. Please check your connection and try again."
      );
    } finally {
      setLoading(false);
    }
  };

  // Helper function to handle API errors
  const handleApiError = (error: any) => {
    if (error.response) {
      // Handle 400 error from our category/difficulty validation
      if (error.response.status === 400) {
        const errorMessage =
          error.response.data?.message ||
          "You already have an active exercise in this category with the same difficulty level.";
        Alert.alert("Cannot Add Exercise", errorMessage);
      } else {
        console.error("API request failed:", error);
        Alert.alert(
          "Error",
          "Failed to reactivate the exercise. Please try again."
        );
      }
    } else {
      console.error("API request failed:", error);
      Alert.alert(
        "Error",
        "Failed to reactivate the exercise. Please check your connection and try again."
      );
    }
  };

  return { loading, reactivateExercise };
}
