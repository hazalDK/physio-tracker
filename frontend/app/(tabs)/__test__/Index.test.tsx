import React from "react";
import { render, fireEvent, waitFor } from "@testing-library/react-native";
import { NavigationContainer } from "@react-navigation/native";
import Index from "../index";
import { useFetchDashboardExercises } from "@/hooks/useFetchDashBoardExercises";
import { useReactivateExercise } from "@/hooks/useReactivateExercise";

// Mock the dependencies
jest.mock("@/hooks/useFetchDashBoardExercises");
jest.mock("@/hooks/useReactivateExercise");

// Mock navigation
const mockNavigate = jest.fn();
jest.mock("@react-navigation/native", () => {
  const actualNav = jest.requireActual("@react-navigation/native");
  return {
    ...actualNav,
    useNavigation: () => ({
      navigate: mockNavigate,
    }),
    // Make useFocusEffect a no-op in tests to prevent infinite re-renders
    useFocusEffect: jest.fn(),
  };
});

// Mocks the react-native-progress Bar component
jest.mock("react-native-progress", () => ({
  Bar: () => "ProgressBar",
}));
jest.mock(
  "@expo/vector-icons/MaterialCommunityIcons",
  () => "MaterialCommunityIcons"
);

describe("Index Component", () => {
  const mockUserExercises = [
    { id: 1, exercise: 101, completed: true, is_active: true },
    { id: 2, exercise: 102, completed: false, is_active: true },
  ];

  const mockExercises = [
    {
      id: 101,
      category: "Knee",
      name: "Exercise 1",
      difficulty_level: "Easy",
      video_id: "abc123",
    },
    {
      id: 102,
      category: "Squat",
      name: "Exercise 2",
      difficulty_level: "Medium",
      video_id: "def456",
    },
  ];

  const mockInactiveExercises = [
    {
      id: 103,
      category: "Sit to stand",
      name: "Exercise 3",
      difficulty_level: "Hard",
      video_id: "ghi789",
      is_active: false,
    },
  ];

  const mockFetchData = jest.fn();
  const mockRefreshData = jest.fn();
  const mockReactivateExercise = jest.fn().mockResolvedValue({});

  beforeEach(() => {
    jest.clearAllMocks();

    (useFetchDashboardExercises as jest.Mock).mockReturnValue({
      loading: false,
      userExercises: mockUserExercises,
      exercises: mockExercises,
      inActiveExercises: mockInactiveExercises,
      fetchData: mockFetchData,
      refreshData: mockRefreshData,
    });

    (useReactivateExercise as jest.Mock).mockReturnValue({
      reactivateExercise: mockReactivateExercise,
    });
  });

  it("renders correctly with exercises", () => {
    const { getByText, getAllByText } = render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    expect(getByText("Completed 50%")).toBeTruthy();
    expect(getByText("Exercise 1")).toBeTruthy();
    expect(getByText("Exercise 2")).toBeTruthy();
    expect(getAllByText(/Difficulty level:/i).length).toBe(2);
  });

  it("shows loading indicator when loading", () => {
    (useFetchDashboardExercises as jest.Mock).mockReturnValue({
      loading: true,
      userExercises: [],
      exercises: [],
      inActiveExercises: [],
      fetchData: mockFetchData,
      refreshData: mockRefreshData,
    });

    const { getByTestId } = render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    expect(getByTestId("loading-indicator")).toBeTruthy();
  });

  it("fetches data on initial render", () => {
    render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    expect(mockFetchData).toHaveBeenCalledTimes(1);
  });

  it("refreshes data when screen gains focus", () => {
    // Extract the callback function when useFocusEffect is called
    let focusCallback: (() => void) | undefined;

    (
      require("@react-navigation/native").useFocusEffect as jest.Mock
    ).mockImplementation((callback) => {
      focusCallback = callback;
    });

    render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    // Reset refreshData call count (from component mount)
    mockRefreshData.mockClear();

    // Execute the focus callback manually
    if (focusCallback) {
      focusCallback();
    }

    expect(mockRefreshData).toHaveBeenCalledTimes(1);
  });

  it("navigates to Exercise screen with correct params when an exercise is pressed", () => {
    const { getByText } = render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    fireEvent.press(getByText("Exercise 1"));

    expect(mockNavigate).toHaveBeenCalledWith("exercise", {
      exerciseId: 101,
      userExerciseId: 1,
    });
  });

  it("navigates to Exercise screen with null userExerciseId when no user exercise exists", () => {
    const exerciseWithoutUserExercise = {
      id: 104,
      category: "Balance",
      name: "New Exercise",
      difficulty_level: "Advanced",
      video_id: "xyz789",
    };

    (useFetchDashboardExercises as jest.Mock).mockReturnValue({
      loading: false,
      userExercises: mockUserExercises,
      exercises: [...mockExercises, exerciseWithoutUserExercise],
      inActiveExercises: mockInactiveExercises,
      fetchData: mockFetchData,
      refreshData: mockRefreshData,
    });

    const { getByText } = render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    fireEvent.press(getByText("New Exercise"));

    expect(mockNavigate).toHaveBeenCalledWith("exercise", {
      exerciseId: 104,
      userExerciseId: null,
    });
  });

  it("navigates to Chatbot screen when the chatbot button is pressed", () => {
    const { getByTestId } = render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    fireEvent.press(getByTestId("chatbot-button"));

    expect(mockNavigate).toHaveBeenCalledWith("chatbot");
  });

  it("opens modal when add button is pressed", () => {
    const { getByTestId, getByText } = render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    fireEvent.press(getByTestId("add-exercise-button"));

    expect(getByText("Add an Exercise")).toBeTruthy();
    expect(getByText("Exercise 3")).toBeTruthy();
  });

  it("reactivates exercise when selected from modal", async () => {
    // Update the mock to include the relationship for inactive exercise
    const updatedUserExercises = [
      ...mockUserExercises,
      { id: 3, exercise: 103, completed: false, is_active: false }, // Add inactive exercise relation
    ];

    (useFetchDashboardExercises as jest.Mock).mockReturnValue({
      loading: false,
      userExercises: updatedUserExercises,
      exercises: mockExercises,
      inActiveExercises: mockInactiveExercises,
      fetchData: mockFetchData,
      refreshData: mockRefreshData,
    });

    const { getByTestId, getByText } = render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    fireEvent.press(getByTestId("add-exercise-button"));
    fireEvent.press(getByText("Exercise 3"));

    await waitFor(() => {
      expect(mockReactivateExercise).toHaveBeenCalledWith(3); // Expect to be called with id 3
      expect(mockRefreshData).toHaveBeenCalled();
    });
  });

  it("closes modal when cancel is pressed", () => {
    const { getByTestId, getByText, queryByText } = render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    fireEvent.press(getByTestId("add-exercise-button"));
    fireEvent.press(getByText("Cancel"));

    expect(queryByText("Add an Exercise")).toBeNull();
  });

  it("handles image error by displaying placeholder image", () => {
    const { getAllByTestId } = render(
      <NavigationContainer>
        <Index />
      </NavigationContainer>
    );

    // Find all Image components by test ID
    const images = getAllByTestId("exercise-image");
    expect(images.length).toBeGreaterThan(0);

    // Simulate image loading error on the first image
    fireEvent(images[0], "error");

    // We can verify the component didn't crash
    expect(images.length).toBeGreaterThan(0);
  });
});
