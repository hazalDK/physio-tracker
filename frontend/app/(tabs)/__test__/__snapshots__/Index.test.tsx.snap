// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`Index Component renders correctly with data 1`] = `null`;
